### PR TITLE
Optimized creation of repeated_state by using expand instead of repeat_interleave.

### PR DESCRIPTION
**Pull request by hoangcuong2011**
Replacing Pytorch's repeat_interleave function with a customized function repeat_interleave_with_expand reduces 48% latency on GPU (K80) and 41% latency on CPU on the creation of repeated_state. This could be worthy to consider given that `RepeatStates` is called numerous times during inference.

Comparison Code:
```
import timeit
set_up_line = """
import torch as pt
pt.manual_seed(1)
beam_size = 5
batch_size = 32
a = pt.rand(7, 2, 1024)
def repeat_interleave_with_expand(state, num_of_repeat, dim):
    state_size = list(state.size())
    expand_state_size = [-1 for _ in range(len(state_size) + 1)]
    state_size[dim] = state_size[dim] * num_of_repeat
    expand_state_size[dim +1] = num_of_repeat
    return state.unsqueeze(dim + 1).expand(expand_state_size).reshape(state_size)
"""
executive_line = """
a1 = repeat_interleave_with_expand(a, beam_size, 0)
a2 = repeat_interleave_with_expand(a, beam_size, 1)
"""
timeit.timeit(executive_line,setup=set_up_line, number=10000)
#gpu: 0.5507124290015781
#cpu: 0.5074962430226151

executive_line = """
a1 = a.repeat_interleave(beam_size, dim=0)
a2 = a.repeat_interleave(beam_size, dim=1)
"""
timeit.timeit(executive_line,setup=set_up_line, number=10000)
#gpu: 1.048011115999543
#cpu: 0.8592731970129535
```
<del>Side note: I still don't know yet why I have to change the code (decoder.py) from 
```source_mask_view = source_mask.view(batch * heads, target_max_len, source_max_len)```
to
```source_mask_view = source_mask.reshape(batch * heads, target_max_len, source_max_len)```
in order to pass `test/integration/test_backwards_compatibility.py` during running `pytest`.</del>

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/decoder.py [hunk 1]
```diff
@@ -251,7 +251,7 @@ def forward(self, step_input: pt.Tensor, states: List[pt.Tensor]) -> Tuple[pt.Te
             autoregr_states = [list(islice(states_iter, 0, layer.num_state_tensors)) for layer in self.layers]  # type: ignore
 
         batch, heads, target_max_len, source_max_len = source_mask.size()
-        source_mask_view = source_mask.view(batch * heads, target_max_len, source_max_len)
+        source_mask_view = source_mask.reshape(batch * heads, target_max_len, source_max_len)
```
Comment (mjdenkowski) [inline sockeye/decoder.py:254]: Have you measured the effect of using `reshape()` instead of `view()` on CPU/GPU latency?

If we make this change, it would also make sense to rename the variable to `source_mask_reshape`.

**Review by mjdenkowski — COMMENTED** [review]
Thanks for working on this!

File: sockeye/decoder.py (hunk 1 above)
Comment (hoangcuong2011) [reply sockeye/decoder.py]: Thanks Michael! I just changed the variable to `source_mask_reshape` as suggested.

Regarding speed, using reshape instead of view increases the latency to `4%` on GPU (K80) and `11%` on CPU.
Note that this reshaping is called only `1` time for each batch. Meanwhile repeat_interleave_with_expand is called a lot more than that (I noticed it is around `50` times for a batch I tried).
Code comparison:

```
import timeit
set_up_line = """
import torch as pt
pt.manual_seed(1)
heads = 8
batch = 20
target_max_len = source_max_len = 50
a = pt.rand(20, 8, 50, 50)
"""
executive_line = """
b = a.reshape(batch * heads, target_max_len, source_max_len)
"""
timeit.timeit(executive_line,setup=set_up_line, number=10000)
#cpu: 0.031618487962987274
#gpu: 0.013913217990193516

executive_line = """
b = a.view(batch * heads, target_max_len, source_max_len)
"""
timeit.timeit(executive_line,setup=set_up_line, number=10000)
#cpu: 0.028475572005845606
#gpu: 0.013381681987084448
```

**Comment by hoangcuong2011** [comment]
Side note: 

`Source_mask` is one of the states (i.e. `steps, source_mask, *other = states` in `TransformerDecoder`). It is thus not possible to run `view` of `source_mask` when it is created by using `expand` instead of `repeat_interleave`. As a result we need to use `reshape` instead of `view` when reshaping this `source_mask` tensor in `TransformerDecoder` class (Luckily this is called only 1 time for each batch).

A way to make it possible to run `view` of `source_mask` is to return a clone of expanding tensor:

`return state.unsqueeze(dim + 1).expand(expand_state_size).clone().reshape(state_size)
`

instead of

`return state.unsqueeze(dim + 1).expand(expand_state_size).clone().reshape(state_size)
`

However, having additional `clone()` increases the latency of the customized function `repeat_interleave_with_expand` by 10% on GPU and 16% on CPU every time this function is called. Given that the customized function is called many more times (e.g. 4x, 5x) than reshaping a `source_mask` tensor, it is better to not include `clone()` in the customized function and use `reshape` instead of `view` of `source_mask` in the `TransformerDecoder` class.

**Comment by hoangcuong2011** [comment]
Thanks Michael! The pull request is closed now do to a lack of evidence whether this improves latency for running `sockeye-translate` with full scale models. 

I still wonder why it does not improve latency because it shows improvement with examples I mentioned above. I suspected it is because of `jit trace`: with `jit trace` the difference between using two functions are minor regarding latency. 

To test this, I simplified the code even further with this [commit](https://github.com/hoangcuong2011/sockeye/commit/6b4dafd83cb0757131db9eb593d9ccd397a66969): https://github.com/hoangcuong2011/sockeye/commit/6b4dafd83cb0757131db9eb593d9ccd397a66969 With this commit, there is no need for `reshape` of `source_mask` in `TransformerDecoder`, as well as there is no need to have explicit additional lists in `repeat_interleave_with_expand` function, i.e. using:

```
state_size = state.size()
return state.unsqueeze(dim + 1).expand(*state_size[:dim + 1], repeats, *state_size[dim + 1:])\
           .clone().view(*state_size[:dim], state.size(dim) * repeats, *state_size[dim + 1:])
```

instead of

```
repeat_state_size = list(state.size())  # type: List[int]
repeat_state_size[dim] = repeat_state_size[dim] * repeats
expand_size = [-1 for _ in range(len(repeat_state_size) + 1)]  # type: List[int]
expand_size[dim + 1] = repeats
return state.unsqueeze(dim + 1).expand(expand_size).reshape(repeat_state_size)
```

@mjdenkowski : Could you later check whether with this commit, it has any latency improvements for running sockeye-translate with full scale models or not? That could be really helpful. Thanks!
